libnopegl/src/pgcraft.c

```diff
@@ -1161,6 +1181,51 @@ static int filter_pipeline_elems(struct pgcraft *s,
     return 0;
 }
 
+static int has_attribute(const struct pgcraft *s, int location)
+{
+    const struct darray *info_array = &s->filtered_pipeline_info.desc.attributes;
+    const struct pipeline_attribute_desc *descs = ngli_darray_data(info_array);
+    for (size_t i = 0; i < ngli_darray_count(info_array); i++) {
+        const struct pipeline_attribute_desc *desc = &descs[i];
+        if (desc->location == location)
+            return 1;
+    }
+    return 0;
+}
+
+static int filter_pipeline_vertex_buffers(struct pgcraft *s)
+{
+    struct pgcraft_pipeline_info *info  = &s->pipeline_info;
+    struct pgcraft_pipeline_info *finfo = &s->filtered_pipeline_info;
+
+    const struct vertex_buffer_layout *layouts = ngli_darray_data(&info->desc.vertex_buffers);
+    const struct buffer **buffers = ngli_darray_data(&info->data.vertex_buffers);
+
+    for (size_t i = 0; i < ngli_darray_count(&info->desc.vertex_buffers); i++) {
+        const struct vertex_buffer_layout *layout = &layouts[i];
+        const struct buffer *buffer = buffers[i];
+        struct vertex_buffer_layout filtered_layout = {
+            .rate = layout->rate,
+            .stride = layout->stride,
+        };
+        for (size_t j = 0; j < layout->nb_attributes; j++) {
+            const struct vertex_attribute *attribute = &layout->attributes[j];
+            if (has_attribute(s, attribute->location)) {
+                filtered_layout.attributes[filtered_layout.nb_attributes++] = *attribute;
```

Comment:
Is there a `nb_attributes` limit check somewhere to prevent the overflow?

Reply:
The only place where the user can set the attributes he wants is in node_render where we have the following code:
``` c
    const size_t nb_attributes = o->attributes ? ngli_hmap_count(o->attributes) : 0;
    if (nb_attributes > limits->max_vertex_attributes) {
        LOG(ERROR, "number of attributes (%zu) exceeds device limits (%u)",
            nb_attributes, limits->max_vertex_attributes);
        return NGL_ERROR_GRAPHICS_LIMIT_EXCEEDED;
    }
```

Reply:
Morever the nb_attributes is also checked in pgcraft:
``` c

        ngli_assert(vertex_buffer.nb_attributes < NGLI_MAX_ATTRIBUTES_PER_BUFFER);
        vertex_buffer.attributes[vertex_buffer.nb_attributes++] = (struct vertex_attribute) {
            .location = base_location + i,
            .format = attribute->format,
            .offset = attribute->offset + i * attribute_offset,
        };
```